File: benchmark.py
#!/usr/bin/env python3
"""
Siena Chat Latency Benchmark — Simple Modern (simplemodern.com)
===============================================================
Measures end-to-end response latency for Siena's AI shopping assistant
across a 10-turn commercial conversation.

METHODOLOGY
-----------
The benchmark drives the Siena chat widget via its internal JavaScript API
injected into the chat iframe (chat.siena.cx). Each "turn" measures:

  t0 = postMessage sent to iframe (action="sendMessage")
  t1 = first new bot message appears in GET /v1/live_chat/message poll

  latency = t1 - t0

This matches real-user perception: time from pressing Send to seeing a response.

API ENDPOINTS DISCOVERED
------------------------
Base: https://api-prod.siena.cx

  POST /v1/live_chat/session?key=<appKey>           → create session
  POST /v1/live_chat/messages                        → submit user message
  POST /v1/live_chat/delivery/confirm                → confirm dedup (message_hash)
  POST /v1/live_chat/track/customer_interaction      → analytics event
  GET  /v1/live_chat/message?page=1&page_size=20     → poll for messages

AUTHENTICATION HEADERS
----------------------
  x-siena-app-key:                  <appKey>
  x-siena-livechat-app-key:         <appKey>
  x-siena-organization-uuid:        <orgUuid>
  x-siena-livechat-integration-uuid: <integrationUuid>
  x-siena-livechat-session-uuid:    <sessionUuid>      (from POST /session)
  x-siena-customer-uuid:            <customerUuid>     (from POST /messages response)
  x-siena-livechat-customer-uuid:   <customerUuid>

NOTE: The GET /v1/live_chat/message endpoint requires an active session
established via the full init flow (postMessage from parent widget). Direct
Python HTTP calls return empty messages because the delivery/confirm
message_hash (MD5-based, computation undetermined) is required to trigger
AI processing. The JavaScript runner below bypasses this by using the
actual widget code path.

JAVASCRIPT RUNNER (paste in browser console on chat.siena.cx tab)
------------------------------------------------------------------
See benchmark.js for the standalone browser runner used to collect results.

RESULTS — 2026-06-29
--------------------
Target    : Simple Modern — Mesa Loop 30oz product page
Session   : 8b169bf7-79a3-4d27-abcc-2c766df79a11
Turns     : 10 (9 valid, 1 anomaly excluded)

  Turn 01:   3727ms  ⚠ (stale bot message, excluded)
  Turn 02:  10395ms
  Turn 03:   9730ms
  Turn 04:  10610ms
  Turn 05:   9989ms
  Turn 06:   6996ms
  Turn 07:  11256ms
  Turn 08:   9248ms
  Turn 09:  10074ms
  Turn 10:   9993ms

  Average latency : 9810ms  (9.8s)
  Median  (p50)   : 9993ms  (10.0s)
  p90             : 11256ms (11.3s)
  Min / Max       : 6996ms / 11256ms
"""

# ── Raw results ────────────────────────────────────────────────────────────────
RESULTS = [
    {"turn": 1,  "message": "What colors does the Mesa Loop 30oz come in?",        "latency_ms": 3727,  "valid": False, "note": "stale bot message"},
    {"turn": 2,  "message": "Does it fit in a standard car cup holder?",           "latency_ms": 10395, "valid": True},
    {"turn": 3,  "message": "Is the straw removable and dishwasher safe?",         "latency_ms": 9730,  "valid": True},
    {"turn": 4,  "message": "What is the difference between the 30oz and 40oz?",   "latency_ms": 10610, "valid": True},
    {"turn": 5,  "message": "Do you offer free shipping on orders?",               "latency_ms": 9989,  "valid": True},
    {"turn": 6,  "message": "How long does standard shipping usually take?",       "latency_ms": 6996,  "valid": True},
    {"turn": 7,  "message": "Can I return it if I do not like the color?",         "latency_ms": 11256, "valid": True},
    {"turn": 8,  "message": "Is this bottle completely leak-proof when closed?",   "latency_ms": 9248,  "valid": True},
    {"turn": 9,  "message": "Does it keep drinks cold for 24 hours?",              "latency_ms": 10074, "valid": True},
    {"turn": 10, "message": "Is the Almond Birch color currently in stock?",       "latency_ms": 9993,  "valid": True},
]


def summarize():
    valid = [r["latency_ms"] for r in RESULTS if r["valid"]]
    s = sorted(valid)
    avg = sum(valid) / len(valid)
    p50 = s[len(s) // 2]
    p90 = s[int(len(s) * 0.9)]

    print("=" * 60)
    print("Siena Chat Latency — Simple Modern (2026-06-29)")
    print("=" * 60)
    for r in RESULTS:
        flag = " ⚠ excluded" if not r["valid"] else ""
        print(f"  Turn {r['turn']:02d}: {r['latency_ms']:>6}ms{flag}")
    print()
    print(f"  Valid turns     : {len(valid)}/10")
    print(f"  Average latency : {avg:.0f}ms  ({avg/1000:.1f}s)")
    print(f"  Median  (p50)   : {p50}ms  ({p50/1000:.1f}s)")
    print(f"  p90             : {p90}ms  ({p90/1000:.1f}s)")
    print(f"  Min / Max       : {min(valid)}ms / {max(valid)}ms")
    print("=" * 60)


if __name__ == "__main__":
    summarize()

File: benchmark.js
/**
 * Siena Chat Latency Benchmark — Browser Runner
 *
 * HOW TO RUN:
 *   1. Open https://www.simplemodern.com/products/mesa-loop-30oz-49
 *   2. Wait for the page to fully load (Siena widget loads automatically)
 *   3. Open DevTools → Console on the PARENT page
 *   4. Paste this entire script and press Enter
 *   5. Results appear in the console and in window._benchResults
 *
 * WHAT IT MEASURES:
 *   t0 = postMessage("sendMessage") dispatched to Siena iframe
 *   t1 = new bot message detected in GET /v1/live_chat/message polling
 *   latency = t1 - t0  (end-to-end, as perceived by the user)
 */

(async function sienaBenchmark() {
  const instance = window.SienaLiveChatWidget?.getInstance?.();
  if (!instance) {
    console.error("Siena widget not found. Make sure you're on the Simple Modern page.");
    return;
  }

  const iframe = document.getElementById("SIENA_CHAT_IFRAME");
  if (!iframe) {
    console.error("Siena iframe not found. Open the chat widget first.");
    return;
  }

  // Extract live session state from the widget instance
  // (these are populated after the widget initializes)
  const getSession = () => {
    const inst = window.SienaLiveChatWidget.getInstance();
    return {
      sessionUuid: inst.currentSessionUuid,
      customerUuid: inst.persistedUserId || inst.currentUserId,
      appKey: inst.apiKeys?.token || inst.config?.APP_KEY,
      integrationUuid: inst.chatSettings?.uuid,
      orgUuid: inst.chatSettings?.organization_uuid,
    };
  };

  const buildHeaders = (sess) => ({
    "Content-Type": "application/json",
    "x-siena-app-key": sess.appKey,
    "x-siena-livechat-app-key": sess.appKey,
    "x-siena-organization-uuid": sess.orgUuid,
    "x-siena-livechat-integration-uuid": sess.integrationUuid,
    "x-siena-livechat-session-uuid": sess.sessionUuid,
    "x-siena-customer-uuid": sess.customerUuid,
    "x-siena-livechat-customer-uuid": sess.customerUuid,
  });

  const API_BASE      = "https://api-prod.siena.cx";
  const POLL_INTERVAL = 500;
  const POLL_TIMEOUT  = 30000;

  const MESSAGES = [
    "What colors does the Mesa Loop 30oz come in?",
    "Does it fit in a standard car cup holder?",
    "Is the straw removable and dishwasher safe?",
    "What is the difference between the 30oz and the 40oz?",
    "Do you offer free shipping on orders?",
    "How long does standard shipping usually take?",
    "Can I return it if I do not like the color?",
    "Is this bottle completely leak-proof when closed?",
    "Does it keep drinks cold for 24 hours?",
    "Is the Almond Birch color currently in stock?",
  ];

  // Ensure chat is open so the widget iframe is active
  window.SienaLaunchChat?.();
  await new Promise(r => setTimeout(r, 1500));

  const sess = getSession();
  console.log(`%c=== Siena Latency Benchmark — 10 turns ===`, "font-weight:bold;color:#0066cc");
  console.log(`Session: ${sess.sessionUuid}`);

  const results = [];

  for (let i = 0; i < MESSAGES.length; i++) {
    const msg = MESSAGES[i];
    const headers = buildHeaders(getSession());

    // Snapshot latest bot message before sending
    const r0 = await fetch(`${API_BASE}/v1/live_chat/message?page=1&page_size=5`, { headers });
    const d0 = await r0.json();
    const botMessages = d0.data.messages.filter(m => m.user?.uuid !== sess.customerUuid);
    const prevBotUuid = botMessages[0]?.uuid || null;

    // Fire the message
    const t0 = performance.now();
    iframe.contentWindow.postMessage({
      action: "sendMessage",
      message: msg,
      pre_generated_question_uuid: null,
      traceContext: { traceId: `bench-${i}`, sessionId: sess.sessionUuid, source: "parent_widget" },
    }, "*");

    // Poll until a NEW bot message appears
    let latency = null;
    const deadline = Date.now() + POLL_TIMEOUT;
    while (Date.now() < deadline) {
      await new Promise(r => setTimeout(r, POLL_INTERVAL));
      const r1 = await fetch(`${API_BASE}/v1/live_chat/message?page=1&page_size=5`, { headers: buildHeaders(getSession()) });
      const d1 = await r1.json();
      const newBotMsgs = d1.data.messages.filter(m => m.user?.uuid !== sess.customerUuid);
      const latest = newBotMsgs[0];
      if (latest && latest.uuid !== prevBotUuid && latest.content) {
        latency = Math.round(performance.now() - t0);
        results.push({ turn: i + 1, message: msg, latency_ms: latency, response: latest.content.slice(0, 80) });
        console.log(`Turn ${i + 1}: %c${latency}ms`, "color:green;font-weight:bold", `| ${msg.slice(0, 50)}`);
        break;
      }
    }

    if (latency === null) {
      results.push({ turn: i + 1, message: msg, latency_ms: null, error: "timeout" });
      console.log(`Turn ${i + 1}: %cTIMEOUT`, "color:red;font-weight:bold", `| ${msg.slice(0, 50)}`);
    }

    // Pause between turns
    await new Promise(r => setTimeout(r, 1500));
  }

  // Summary
  const valid = results.filter(r => r.latency_ms !== null).map(r => r.latency_ms);
  const sorted = [...valid].sort((a, b) => a - b);
  const avg = Math.round(valid.reduce((a, b) => a + b, 0) / valid.length);
  const p50 = sorted[Math.floor(sorted.length * 0.5)];
  const p90 = sorted[Math.floor(sorted.length * 0.9)];

  console.log(`%c\n=== RESULTS ===`, "font-weight:bold;color:#0066cc");
  console.table(results.map(r => ({
    Turn: r.turn,
    "Latency (ms)": r.latency_ms ?? "TIMEOUT",
    Question: r.message.slice(0, 50),
  })));
  console.log(`%cAverage: ${avg}ms | p50: ${p50}ms | p90: ${p90}ms | Min: ${sorted[0]}ms | Max: ${sorted[sorted.length - 1]}ms`,
    "font-weight:bold");

  window._benchResults = results;
  window._benchSummary = { avg_ms: avg, p50_ms: p50, p90_ms: p90, min_ms: sorted[0], max_ms: sorted[sorted.length - 1], n_valid: valid.length };

  return window._benchSummary;
})();
